pkg/jsonpb/ttn.go

```diff
@@ -14,25 +14,124 @@
 
 package jsonpb
 
-import "github.com/grpc-ecosystem/grpc-gateway/runtime"
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"os"
+	"strconv"
+
+	"github.com/TheThingsIndustries/protoc-gen-go-json/jsonplugin"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+)
+
+var jsonpluginFeatureFlag = func() bool {
+	b, _ := strconv.ParseBool(os.Getenv("TTN_LW_EXP_JSONPLUGIN"))
+	return b
+}()
 
 // TTN returns the default TTN JSONPb marshaler.
-func TTN() *GoGoJSONPb {
-	return &GoGoJSONPb{
-		OrigName:    true,
-		EnumsAsInts: true,
+func TTN() *TTNMarshaler {
+	return &TTNMarshaler{
+		GoGoJSONPb: &GoGoJSONPb{
+			OrigName:    true,
+			EnumsAsInts: true,
+		},
+	}
+}
+
+type TTNMarshaler struct {
+	*GoGoJSONPb
+}
+
+func (*TTNMarshaler) ContentType() string { return "application/json" }
+
+func (m *TTNMarshaler) Marshal(v interface{}) ([]byte, error) {
+	if marshaler, ok := v.(jsonplugin.Marshaler); ok && jsonpluginFeatureFlag {
```

Comment:
Nit: I would wrap everything with the feature flag, to avoid the cast completely. Same bellow.